functions/executor/call.py

```diff
@@ -104,33 +111,183 @@ def exec_dag_function(pusher_cache, kvs, triggers, function, schedule):
+            address_to_versioned_key_list_map[addr] = list(trigger.address_to_versioned_key_list_map[addr].versioned_keys)
+        # combine dependency from previous func
+        for dep in trigger.dependency:
+            if dep.key in dependency:
+                dependency[dep.key] = _merge_vector_clock(dependency[dep.key], dep.vector_clock)
+            else:
+                dependency[dep.key] = dep.vector_clock
+
+    logging.info('Executing function %s for DAG %s (ID %d) in causal consistency.' %
+            (schedule.dag.name, fname, trigger.id))
+
+    fargs = _process_args(fargs)
+
+    kv_pairs = {}
+
+    result = _exec_func_causal(kvs, function, fargs, kv_pairs, schedule, address_to_versioned_key_list_map)
+
+    for key in kv_pairs:
+        if key in dependency:
+            dependency[key] = _merge_vector_clock(dependency[key], kv_pairs[key][0])
+        else:
+            dependency[key] = kv_pairs[key][0]
+
+    is_sink = True
+    for conn in schedule.dag.connections:
+        if conn.source == fname:
+            is_sink = False
+            new_trigger = DagTrigger()
+            new_trigger.id = trigger.id
+            new_trigger.target_function = conn.sink
+            new_trigger.source = fname
+
+            if type(result) != tuple:
+                result = (result,)
+
+            al = new_trigger.arguments
+            al.args.extend(list(map(lambda v: serialize_val(v, None, False),
+                result)))
+
+            new_trigger.address_to_versioned_key_list_map = address_to_versioned_key_list_map
+
+            for key in dependency:
+                dep = new_trigger.dependency.add()
+                dep.key = key
+                dep.vector_clock = dependency[key]
+
+            dest_ip = schedule.locations[conn.sink]
+            sckt = pusher_cache.get(sutils._get_dag_trigger_address(dest_ip))
+            sckt.send(new_trigger.SerializeToString())
+
+    if is_sink:
+        logging.info('DAG %s (ID %d) completed in causal mode; result at %s.' %
+                (schedule.dag.name, trigger.id, schedule.response_id))
+
+        vector_clock = {}
+        if schedule.response_id in dependency:
+            if schedule.id in dependency[schedule.response_id]:
+                dependency[schedule.response_id][schedule.id] += 1
+            else:
+                dependency[schedule.response_id][schedule.id] = 1
+            vector_clock = dependency[schedule.response_id][schedule.id]
+            del dependency[schedule.response_id]
+        else:
+            vector_clock = {schedule.id : 1}
+
+        succeed = kvs.causal_put(schedule.response_id, vector_clock, dependency, serialize_val(result), schedule.id)
+        while not succeed:
+            kvs.causal_put(schedule.response_id, vector_clock, dependency, serialize_val(result), schedule.id)
+
+
+
+def _exec_func_causal(kvs, func, args, kv_pairs, schedule, address_to_versioned_key_list_map):
     func_args = ()
 
+    to_resolve = []
```

Comment:
Same point here as with `resolve_ref_normal`.

Reply:
as mentioned before, I'm packing them to one request..